File: services/auth-service/lib/gmail/notify.js
const { claimDeadGmailNotice, unclaimDeadGmailNotice } = require('../crypto/migrate');
const { sendDiscordDM, deadGmailMessage } = require('../discord');

async function notifyDeadGmail(pool, userId, email) {
    const claimed = await claimDeadGmailNotice(pool, userId, email);
    if (!claimed) return;
    if (!claimed.discord_id) {
        console.log(`dead gmail noticed mailbox=${claimed.email} (no discord; skip dm)`);
        return;
    }
    try {
        await sendDiscordDM(claimed.discord_id, deadGmailMessage(claimed.email));
        console.log(`dead gmail discord dm sent mailbox=${claimed.email}`);
    } catch (err) {
        console.error(`dead gmail discord dm failed for ${claimed.email}: ${err.message}`);
        await unclaimDeadGmailNotice(pool, userId, claimed.email);
    }
}

module.exports = { notifyDeadGmail };

File: services/ingestion-service/internal/notify/notify.go
package notify

import (
	"context"
	"database/sql"
	"log"
	"strings"

	"sift/ingestion/internal/discorddm"
)

func DeadGmail(ctx context.Context, db *sql.DB, userID, email string) {
	email = strings.TrimSpace(email)
	if userID == "" || email == "" {
		return
	}

	var discordID, mailbox string
	err := db.QueryRowContext(ctx, `
		UPDATE oauth_credentials AS c
		SET token_invalid_notified_at = NOW()
		FROM users u
		WHERE c.user_id = u.id
		  AND c.user_id = $1
		  AND c.provider = 'google'
		  AND lower(c.email) = lower($2)
		  AND c.token_invalid_notified_at IS NULL
		RETURNING u.discord_id, c.email
	`, userID, email).Scan(&discordID, &mailbox)
	if err == sql.ErrNoRows {
		return
	}
	if err != nil {
		log.Printf("dead gmail notice claim failed for %s: %v", email, err)
		return
	}

	if err := discorddm.Send(ctx, discordID, discorddm.DeadGmailMessage(mailbox)); err != nil {
		log.Printf("dead gmail discord dm failed for %s: %v", mailbox, err)
		if _, resetErr := db.ExecContext(ctx, `
			UPDATE oauth_credentials
			SET token_invalid_notified_at = NULL
			WHERE user_id = $1 AND provider = 'google' AND lower(email) = lower($2)
		`, userID, mailbox); resetErr != nil {
			log.Printf("dead gmail notice unclaim failed for %s: %v", mailbox, resetErr)
		}
		return
	}
	log.Printf("dead gmail discord dm sent mailbox=%s", mailbox)
}
